File: leetcode/JavaScript/No18.4sum.js
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given an array S of n integers, are there elements a, b, c, and d in S such that a + b + c + d = target?
 * Find all unique quadruplets in the array which gives the sum of target.
 *
 * Note:
 * The solution set must not contain duplicate quadruplets.
 *
 * Example:
 * given array S = [1, 0, -1, 0, -2, 2], and target = 0.
 * A solution set is:
 * [
 *    [-1,  0, 0, 1],
 *    [-2, -1, 1, 2],
 *    [-2,  0, 0, 2]
 * ]
 *
 * 四数求和，类似于之前的三数求和，只是注意处理重复值的问题
 */

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
  nums.sort((n1, n2) => n1 - n2)
  const result = []

  // 利用双指针
  const findSum = (start, end, num, arr) => {
    let i = start
    let j = end
    while (i < j) {
      const sum = nums[i] + nums[j]
      if (sum <= num) {
        if (sum === num) result.push([...arr, nums[i], nums[j]])
        while (i < end && nums[i] === nums[i + 1]) i += 1
        i += 1
      } else {
        while (j > start && nums[j] === nums[j - 1]) j -= 1
        j -= 1
      }
    }
  }

  for (let i = 0; i < nums.length - 3; i += 1) {
    if (nums[i] === nums[i - 1]) continue

    for (let j = nums.length - 1; j > i + 2; j -= 1) {
      if (nums[j] === nums[j + 1]) continue
      const remain = target - nums[i] - nums[j]
      findSum(i + 1, j - 1, remain, [nums[i], nums[j]])
    }
  }
  return result
}

File: leetcode/Ruby/No18.4sum.rb
=begin
Difficulty:
Medium

Desc:
Given an array nums of n integers and an integer target, are there elements a, b, c, and d in nums such that a + b + c + d = target?
Find all unique quadruplets in the array which gives the sum of target.

Note:
The solution set must not contain duplicate quadruplets.

Example:
Given array nums = [1, 0, -1, 0, -2, 2], and target = 0.
A solution set is:
[
  [-1,  0, 0, 1],
  [-2, -1, 1, 2],
  [-2,  0, 0, 2]
]
=end

def three_sum(nums, start, target, pre_num)
  results = []
  index = start

  while index < nums.size - 2 do
    num = nums[index]

    if index > start && num == nums[index -1] then
      index += 1
      next
    end

    i = index + 1
    j = nums.size - 1

    while i < j do
      sum = nums[i] + nums[j] + num
      if sum > target then
        j -= 1
        while i < j && nums[j] == nums[j + 1] do
          j -= 1
        end
      elsif sum < target then
        i += 1
        while i < j && nums[i] == nums[i - 1] do
          i += 1
        end
      else
        results.push([pre_num, num, nums[i], nums[j]])
        i += 1
        j -= 1

        while i < j && nums[i] == nums[i - 1] do
          i += 1
        end
        while i < j && nums[j] == nums[j + 1] do
          j -= 1
        end
      end
    end

    index += 1
  end

  return results
end

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
def four_sum(nums, target)
  nums.sort!
  results = []

  i = 0
  while i < nums.size - 3 do
    num = nums[i]
    if i == 0 || num != nums[i - 1] then
      results += three_sum(nums, i + 1, target - num, num)
    end
    i += 1
  end

  return results
end

# Test case
print four_sum([1, 0, -1, 0, -2, 2], 0)
print four_sum([0,0,0,0], 0)
print four_sum([-3,-2,-1,0,0,1,2,3], 0) # [[-3, -2, 2, 3], [-3, -1, 1, 3], [-3, 0, 0, 3], [-3, 0, 1, 2], [-2, -1, 0, 3], [-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]]
print four_sum([1,-2,-5,-4,-3,3,3,5], -11) # [[-5,-4,-3,1]]
print four_sum([-1,0,1,2,-1,-4], -1)
